### PR TITLE
Add online multiplayer

### PR DESCRIPTION
**Pull request by Clebiez**
# As Brice, I want to play online with a friend

[Trello Card](https://trello.com/c/wq3gd5GZ/2-1-as-brice-i-want-to-play-online-with-a-friend-1)

- [x] When a player create a game, fill player1Hash by his hash.
- [x] Give it a waiting page to invite a friend
- [x] When accessing to game/:id/invite, check if player2 exist in the game, instead fill it by this user.
- [x] If 2 players are in the game, it can started.
- [x] Add yourTurn value to block UI if it's not your turn
- [x] Add witness mode & allow you to see in live the game (no cost to implement it, I know it was another ticket but it was most  expensive to disable UI for this witness than give it the live game.)
- [x] Finally add refresh loop if it's not your turn to get new actions made by other player. 

### When the game is started
multiplayer

### When you are alone on a game & waiting for player 2
image

### CODE REVIEW

File: templates/game/index.html.twig [hunk 1]
```diff
@@ -7,76 +7,25 @@
...
 
-		{# Each cell has a couple of coords as value : "row-column" (e.g : "O-5")
-        Twig inject in loops an object "loop".
-        loop.index is the iterator (start 1 to 6).
-        loop.parent be able to access to the first loop object.
-        #}
+{% block javascripts %}
+	{% if status == constant('App\\Service\\GameService::GAME_WAITING_OPPONENT') or isYourTurn == false %}
+		<script type="text/javascript">
+			window.setTimeout(() => {
```
Comment (zyhou) [inline templates/game/index.html.twig:25]: :+1:

File: templates/base.html.twig [hunk 2]
```diff
@@ -3,6 +3,8 @@
     <head>
         <meta charset="UTF-8">
         <title>{% block title %}Pentago !{% endblock %}</title>
+        {% block metas %}{% endblock %}
```
Comment (zyhou) [inline templates/base.html.twig]: can be delete?

File: src/Controller/GameController.php [hunk 3]
```diff
@@ -41,38 +43,81 @@ public function newGame(): Response
...
+     * @Route("/game/{id}", name="game")
+     * Used to display the board
+     */
+    public function game(Request $request, string $id): Response
+    {
+        $playerHash = $request->cookies->get($this::COOKIE_KEY);
+
+        $entityManager = $this->getDoctrine()->getManager();
+        $game = $entityManager->getRepository(Game::class)->find($id);
+        if (!$game) {
```
Comment (zyhou) [inline src/Controller/GameController.php:108]: You don't test `$playerHash` here ?

File: src/Controller/GameController.php [hunk 4]
```diff
@@ -41,38 +43,81 @@ public function newGame(): Response
...
+            // Set this player as player 1
+            $game->setPlayer1Hash($playerHash);
+        } elseif ($playerHash !== $game->getPlayer1Hash() && $game->getPlayer2Hash() === null) {
+            // If game has no player 2
+            // Start this game !
+            $game = $this->gameService->setPlayer2AndStartGame($playerHash, $game);
+        }
 
-        if (!$playerHash || !$game) {
+        $entityManager->flush();
```
Comment (zyhou) [inline src/Controller/GameController.php]: The spectator is doing to much on this function. You need to generate a hash and flush the entity in spectator case?

**Review by zyhou — CHANGES_REQUESTED** [review]

File: src/Controller/GameController.php (hunk 3 above)
Comment (Clebiez) [reply src/Controller/GameController.php:108]: You can access to a game without playerHash and in this case, you have to be a witness.
You only need one if you want to play.

The following line test actually if $playerHash is in the game (and exist by the way)

`117:$yourValue = $this->gameService->getPlayerValue($game, $playerHash);`

And be able to determine if it's a witness or a player

File: src/Controller/GameController.php [hunk 5]
```diff
@@ -62,21 +62,25 @@ public function inviteGame(Request $request, string $id): Response
             return $this->redirectToRoute('newGame');
         }
 
-        if (!$playerHash) {
-            $playerHash = $this->playerService->generatePlayerHash();
+        // If game is not started yet, we waiting for players !
+        if (!$this->gameService->isStarted($game)) {
```
Comment (zyhou) [inline src/Controller/GameController.php:66]: I like it ;)

**Review by zyhou — APPROVED** [review]

**Pull request merged**